### PR TITLE
CUDA: Only convert autolocals to args if there are local function args

### LINKED ISSUE #832: CUDA regression: SHOC scan gives wrong results
It seems like https://github.com/pocl/pocl/commit/ec3cd592f7e16f8154c1431a4b0b5659ff7b06d1 is to blame. I'm trying to figure out what is going on here.

### PR DESCRIPTION
**Pull request by isuruf**
This PR reverts the previous PR and modifies the AutomaticLocals pass to have an optional behaviour of only converting if there are local function args.

After multiple tries, I couldn't get the automatic locals + local function args to work correctly, and those kernels might have bad performance, but correctness is more important.

Fixes #832

### CODE REVIEW

**Comment by isuruf** [comment]
Performance of GEMM benchmark in SHOC is not affected by this change as it doesn't use local function args.

**Comment by pjaaskel** [comment]
What is the symptom of the breakage? Wrong results or crash?

**Comment by isuruf** [comment]
Wrong results in the case of https://github.com/pocl/pocl/issues/832

Another issue is that if we have both autolocals and local function arguments, we need to make it one extern block, but the optimizer might have demoted the "lllvm global" variable to the function scope and may hvae run other optimizations. Therefore we don't see the autolocal "llvm global" variable when we are generating code.

**Comment by pjaaskel** [comment]
> Another issue is that if we have both autolocals and local function arguments, we need to make it one extern block, but the optimizer might have demoted the "lllvm global" variable to the function scope and may hvae run other optimizations. Therefore we don't see the autolocal "llvm global" variable when we are generating code.

OK, I think you need a new automatic locals handing mode for this. AutomaticLocals::processAutomaticLocals(Function *F) in AutomaticLocals.cc is where it does the conversion from the LLVM global placeholders to function arguments.

File: lib/llvmopencl/AutomaticLocals.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+
+#include "llvm/Pass.h"
+
+namespace pocl {
+// autolocals_to_args can have the following values
+//  0: do not convert autolocals to args
+//  1: convert autolocals to args
+//  2: convert autolocals to args only if there are local function arguments
```
Comment (isuruf) [inline lib/llvmopencl/AutomaticLocals.h:29]: > OK, I think you need a new automatic locals handing mode for this. AutomaticLocals::processAutomaticLocals(Function *F) in AutomaticLocals.cc is where it does the conversion from the LLVM global placeholders to function arguments.

@pjaaskel, that's what I've implemented here.

**Comment by isuruf** [comment]
Ping on this

File: lib/CL/devices/cuda/pocl-cuda.c [hunk 2]
```diff
@@ -151,7 +193,7 @@ pocl_cuda_init (unsigned j, cl_device_id dev, const char *parameters)
   /* TODO: Get images working */
   dev->image_support = CL_FALSE;
 
-  dev->autolocals_to_args = CL_FALSE;
+  dev->autolocals_to_args = 2;
```
Comment (pjaaskel) [inline lib/CL/devices/cuda/pocl-cuda.c:196]: Can you create an enum out of this magic number to improve readability?

File: lib/CL/pocl_llvm_wg.cc [hunk 3]
```diff
@@ -23,11 +23,12 @@
    THE SOFTWARE.
 */
 
+#include "../llvmopencl/AutomaticLocals.h"
```
Comment (pjaaskel) [inline lib/CL/pocl_llvm_wg.cc:26]: Can you add llvmopencl to the include paths of this directory's CMake configuration instead?

File: lib/llvmopencl/AutomaticLocals.cc [hunk 4]
```diff
@@ -54,7 +55,9 @@ namespace {
 
   public:
     static char ID;
-    AutomaticLocals() : ModulePass(ID) {}
+    int autolocals_to_args;
+    AutomaticLocals(int autolocals_to_args = 1)
```
Comment (pjaaskel) [inline lib/llvmopencl/AutomaticLocals.cc:59]: Please use an enum/macro instead of a magic number here.

File: lib/llvmopencl/AutomaticLocals.cc [hunk 5]
```diff
@@ -169,6 +179,21 @@ AutomaticLocals::processAutomaticLocals(Function *F) {
     return F;
   }
 
+  if (autolocals_to_args == 2) {
```
Comment (pjaaskel) [inline lib/llvmopencl/AutomaticLocals.cc:182]: Please use the enum here.

File: lib/llvmopencl/AutomaticLocals.cc [hunk 6]
```diff
@@ -169,6 +179,21 @@ AutomaticLocals::processAutomaticLocals(Function *F) {
...
+    bool NeedsArgOffsets = false;
+    // Loop over arguments.
```
Comment (pjaaskel) [inline lib/llvmopencl/AutomaticLocals.cc:184]: I think this comment is not needed thanks to the great new C++ loop syntax :)

File: lib/llvmopencl/AutomaticLocals.cc [hunk 7]
```diff
@@ -169,6 +179,21 @@ AutomaticLocals::processAutomaticLocals(Function *F) {
...
+    for (auto &Arg : F->args()) {
+      // Check for local memory pointer.
+      llvm::Type *ArgType = Arg.getType();
+      if (ArgType->isPointerTy() && ArgType->getPointerAddressSpace() == 3) {
```
Comment (pjaaskel) [inline lib/llvmopencl/AutomaticLocals.cc:188]: Please use the macro not a magic number here for the local address space id.

File: lib/llvmopencl/AutomaticLocals.h [hunk 8]
```diff
@@ -0,0 +1,31 @@
+// Create a AutomaticLocals pass for converting automatic locals to args
```
Comment (pjaaskel) [inline lib/llvmopencl/AutomaticLocals.h:1]: an

File: lib/llvmopencl/AutomaticLocals.h [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+
+#include "llvm/Pass.h"
+
+namespace pocl {
+// autolocals_to_args can have the following values
+//  0: do not convert autolocals to args
```
Comment (pjaaskel) [inline lib/llvmopencl/AutomaticLocals.h:27]: Replace these comments with an enum with possible additional comment per value.

File: lib/llvmopencl/AutomaticLocals.h [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+//  1: convert autolocals to args
```
Comment (pjaaskel) [inline lib/llvmopencl/AutomaticLocals.h:28]: Perhaps add 'always' to the beginning of this enum/comment.

File: lib/llvmopencl/AutomaticLocals.h (hunk 1 above)
Comment (pjaaskel) [inline lib/llvmopencl/AutomaticLocals.h:29]: ...there are dynamic local memory function arguments in the kernel.

**Comment by isuruf** [comment]
> In addition to the comments above, does this PR add new passing test cases?

No, it fixes a regression from one of my previous PRs.
I'll add the SHOC benchmarks as a tier 1 testsuite (which was where this regression was noticed) in a follow up PR.

File: lib/CL/pocl_cl.h [hunk 11]
```diff
@@ -638,6 +638,17 @@ typedef struct pocl_global_mem_t {
...
+/**
+ * Enumeration for different modes of converting automatic locals
+ */
+typedef enum {
+  POCL_AUTOLOCALS_TO_ARGS_NEVER = 0,
+  POCL_AUTOLOCALS_TO_ARGS_ALWAYS = 1,
+  // convert autolocals to args only if there are dynamic local memory function
+  // arguments in the kernel.
+  POCL_AUTOLOCALS_TO_ARGS_ONLY_IF_DYNAMIC_LOCALS_PRESENT = 2,
+} pocl_autolocals_to_args_type;
```
Comment (pjaaskel) [inline lib/CL/pocl_cl.h:650]: Can you rename to pocl_autolocals_to_args_strategy in your next PR?

**Comment by isuruf** [comment]
Thanks

**Pull request merged**